deploy/templates/gcs-manifests/gcs-gd2.yml.j2

```diff
@@ -1,20 +1,22 @@
 ---
-kind: DaemonSet
+kind: StatefulSet
```

Comment:
From my reading, it looks like it will move to the 4th node, but the state & disks won't.
There will need to be a nodeAffinity stanza added to lock it to the host. This will be necessary when we are directly accessing `/dev` to get our storage w/o kube being aware.

Reply:
I would like to know why a StatefulSet instead of Deployment. My best guess is the predictable pod name (and so DNS name of `{{ kube_hostname }}-0` (or something). But is that actually needed?

When the pod restarts and gets a new IP...
- The Service will automatically update, so the client access will update appropriately
- The pod will find its state directory (pending hostPath fix) and re-register w/ etcd using the old gd2 node UUID. Is this re-registration is sufficient for the peers to find the restarted pod?

Are there any other barriers to using a Deployment?

Reply:
I agree that it seems like Deployments with static Service names should suffice for this purpose. Since you can add a GD2 node by DNS name (e.g. the Service address) and, as mentioned, the UUID is stored in the persistent volume of the Pod, cluster state would be maintained.

Reply:
Just to be clear, I'm suggesting getting rid of stable DNS and IP of the pods. All that would remain is the overall glusterd2 Service that would spread across the entire cluster. This Service would be used both for mounts (volfile fetch) and management ops.
Volfiles would have the pod IPs, but I think we decided these would get regenerated automatically and pushed to clients when the IPs change.

Reply:
If you got rid of the stable identifiers, how would you maintain cluster state? If a node goes down and comes back with the same UUID but different name/IP, will the other nodes be able to find it or will it contact all the other nodes and add its new name/IP to its list of known aliases? Further, if the entire cluster goes down and all nodes change identity, how will they rebuild?

Reply:
GD2 resolves the hostname to get the IP it needs to listen on. I'm not sure if using a normal Service would work, as the resolved address wouldn't be available in the pod. I'll test it out.

About nodeAffinity, isn't specifying a nodeSelector enough? All we need is to ensure that the specific pods are only scheduled, run and restarted on a particular node. We don't need any other constraints provided by nodeAffinity, and nodeSelector just works fine.

Reply:
Also, I just have general question about StatefulSet vs Deployment (or ReplicaSet). Both perform nearly the same functions, with the addition that StatefulSet give us a consistent IDs for free, and Deployments don't. With Deployment we'd need additional Services per deployment to get ourselves consistent names. I don't see any real advantages to using Deployments over StatefulSets.

Reply:
I tested out Deployments + Service with clusterip, it doesn't work. GD2 fails to start as it cannot bind to the service ip which is not available on an interface within the pod.

Reply:
Deployments were developed during a time where consistent Pod identities were not seen as a necessity. The way k8s was designed, an app should have had no need to know a specific Pod's IP, just the Service name, since it cared about the other app's logic not the location. The main goal of a Deployment was to manage Pod scaling and recovery. StatefulSets were brought up later, after enough people complained that they wanted persistent identities for their replicated applications.

If GD2 needs to bind to an IP on the Pod, then my previous question remains: When the IP changes, is the GD2 cluster somehow able to auto-detect the new IP and recover? If not, you should be using `hostNetworking`.

Reply:
> When the IP changes, is the GD2 cluster somehow able to auto-detect the new IP and recover?

@kshlm can correct my understanding here... gd2 registers w/ etcd using its node uuid, and when it does, it updates its IP in etcd also, so the current IP of the gd2 nodes can be read from etcd.

Reply:
> I tested out Deployments + Service with clusterip, it doesn't work. GD2 fails to start as it cannot bind to the service ip which is not available on an interface within the pod.

Is this caused by the introduction of GD2_CLIENTADDRESS/PEERADDRESS? Can't it just bind to all pod IPs, allowing us to use a normal service?

I also went looking a bit at StatefulSet/Deployment as it's been coming up in various places... CSI uses StatefulSet also, and I'm thinking it is the semantic difference in the two... "At most one" vs. "At least one"... SS takes down a given instance and replaces it when updating whereas Deployment may start the new before taking down the old (something we don't want). Based on this, SS seems to be the correct choice.

> About nodeAffinity, isn't specifying a nodeSelector enough? All we need is to ensure that the specific pods are only scheduled, run and restarted on a particular node. We don't need any other constraints provided by nodeAffinity, and nodeSelector just works fine.

Two things...
1. We need to make sure we don't get more than 1 on a node using the same devices. DaemonSet limits to 1 per node... SS doesn't.
2. There is no guarantee all nodes will be using the same set of devices

(and this is something that's going to end up back in the operator discussion...) For any instance that uses `/dev`, there needs to be a mandatory nodeAffinity to lock it to that specific node. `requiredDuringSchedulingIgnoredDuringExecution` with `kubernetes.io/hostname` set to the hostname.

Reply:
> Is this caused by the introduction of GD2_CLIENTADDRESS/PEERADDRESS? Can't it just bind to all pod IPs, allowing us to use a normal service?

I messed around with this somewhat. I think the peer address configuration itself is not sufficient. The configured address must ALSO exist on the node / in the pod in order to be valid. Using the Service IP won't work, since it wouldn't be recognized it as a device in the container.